### PR TITLE
chore: batch billing events

### PR DESCRIPTION
**Pull request by TBonnin**
The batching happens in memory and events are sent on an interval (or when maxBatchSize is reached) in order to avoid sending too many individual events

Tested in staging

---

**Batch Billing Events for Efficient Ingestion and Graceful Shutdown**

This PR introduces an in-memory batching system for billing events via a generic Batcher class. Instead of sending billing events immediately, events are now accumulated and dispatched in configurable batches, either when a batch size threshold is reached or after a time interval, reducing API call volume and improving shutdown safety. The implementation includes queue/retry/discard safeguards, integration across server and persist app shutdown flows, environment-driven configuration, and comprehensive unit tests to cover all paths. All usages of the previous billing send interfaces are updated to the new async batching API.

**Key Changes:**
• Introduces packages/billing/lib/batcher.ts: a generic and configurable batching/queueing/retry class.
• Refactors packages/billing/lib/billing.ts: all event submission now goes through the batcher (add/`addAll`, not send/`sendAll`).
• Adds shutdown handling: ensures batcher queue is flushed/drained on both server and persist shutdown.
• Configures batching parameters (batch size, interval, queue/max retries) via new environment variables.
• All previous immediate billing event calls throughout server, persist, crons now use the batching model.
• Comprehensive unit test coverage for batcher logic (process, flush, retry, limit, timer, shutdown).
• New logging utility for billing domain.
• `DevDependency` update: Vitest for tests.

**Affected Areas:**
• packages/billing/lib/batcher.ts
• packages/billing/lib/billing.ts
• packages/billing/lib/batcher.unit.test.ts
• packages/billing/lib/logger.ts
• packages/billing/package.json
• packages/utils/lib/environment/parse.ts
• packages/persist/lib/app.ts
• packages/persist/lib/records.ts
• packages/server/lib/server.ts
• packages/server/lib/crons/usage.ts
• packages/server/lib/controllers/sync.controller.ts
• package-lock.json

**Potential Impact:**

**Functionality**: Makes billing ingestion asynchronous and buffered, reducing API call frequency and improving shutdown reliability (less risk of lost billing events). Applications expecting immediate billing effects must account for batching delay.

**Performance**: Reduces API/network overhead but introduces slight asynchronous delay for billing event processing. Memory usage is now bounded by the configured batch/queue sizes (default 50k).

**Security**: No direct security impact. Reliable event drain and bounded queues reduce risk of silent data loss/corruption.

**Scalability**: Much improved: system now handles bursts and high event rates with tunable upper bounds and retry/discard policies.

**Review Focus:**
• Thread/concurrency safety in batcher (queue vs timer vs shutdown).
• All legacy billing.send/`sendAll` usage replaced with add/`addAll`.
• Integration of batcher shutdown in all server and persist exit flows.
• Correctness and completeness of batcher unit test logic, especially for edge/retry/timeout.
• Enforcement of configured limits to avoid memory leaks.

<details>
<summary><strong>Testing Needed</strong></summary>

• Verify batching configuration via environment variables.
• End-to-end test with high event volume: ensure all events are counted, including after shutdowns/stopping servers.
• Simulate abrupt process termination to confirm correct retry/discard behavior; observe logs for any lost events.
• Run all billing-related flows to confirm non-regression, especially record persist and action events.

</details>

<details>
<summary><strong>Code Quality Assessment</strong></summary>

**packages/billing/lib/batcher.ts**: Well-structured, highly readable, covers all expected operational states. Comments clarify non-obvious design. Concurrency logic is careful. Queue/retry/interval logic well tested.

**packages/billing/lib/batcher.unit.test.ts**: Comprehensive, isolates all flow/branches, uses fake timers cleanly.

**packages/billing/lib/billing.ts**: Well-documented, clean refactor, clear error paths.

**other files**: Systematic API usage update, minimal risk.

</details>

<details>
<summary><strong>Best Practices</strong></summary>

**Testing**:
• Comprehensive unit coverage for all batcher branches
• Mocked process/tested error handling

**Configuration**:
• All batch behavior tunable via env vars with validation

**Logging**:
• Clear error and success logging on batch/flush/shutdown

**Resource Management**:
• Bounded queues to prevent unbounded memory growth
• No retry storms

**Concurrency**:
• Safe timer management
• Atomic queue/flush/shutdown sequencing

</details>

<details>
<summary><strong>Possible Issues</strong></summary>

• Abrupt process exit (non-graceful shutdown) will drop in-memory queued events.
• In extreme concurrency or rapid-fire shutdown, may miss a final flush-unit tests help but integration testing for exit edge cases is warranted.
• If event rate exceeds ingest capacity, events will be discarded up to the configured limit (with explicit error returned).
• Batches that fail after max retries are dropped and logged-monitor for unexplained event loss.

</details>

---
*This summary was automatically generated by @propel-code-bot*

### CODE REVIEW

File: packages/billing/lib/batcher.ts [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    item: T;
+    retries: number;
+}
+
+export class Batcher<T> {
+    private readonly process: (events: T[]) => Promise<void>;
+    private readonly maxBatchSize: number;
+    private readonly flushInterval: number;
+    private queue: Item<T>[];
+    private readonly maxQueueSize: number;
```
Comment (TBonnin) [inline packages/billing/lib/batcher.ts:14]: in order to have an upper limit and avoid the queue growing unbounded in case orb is down for instance. Set to 50_000 by default.

File: packages/billing/lib/batcher.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    private readonly maxProcessingRetry: number;
```
Comment (TBonnin) [inline packages/billing/lib/batcher.ts:15]: in order to avoid `poison-pill events` that would be retried for ever. Default value is set to 3 retries

**Review by bodinsamuel — APPROVED** [review]
looks good
